### PR TITLE
LPS-115019 Add a "view state" display context holder for search containers to the Clay Taglib

### PR DESCRIPTION
**Pull request by jbalsas**
From https://issues.liferay.com/browse/LPS-115019:

> The Clay Taglib currently contains a [ManagementToolbarDisplayContext](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/display/context/ManagementToolbarDisplayContext.java) interface which has the potential for encapsulating "view state" for seach container toolbar.
> 
> **This task involves introducing a new `com.liferay.frontend.taglib.clay.servlet.taglib.view.state` package with interfaces like `ClayToolBarViewState` which `extends` `ManagementToolbarDisplayContext` and adds some additional methods.**
> 
> It also involves support for encapsulating the view state within a `javax.portlet.RenderURL`. [When combined with a Liferay Friendly URL](https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/resources/META-INF/friendly-url-routes.xml), we will be able to achieve human-readable, bookmarkable friendly portlet RenderURLs that look like the following:
http://localhost:8080/web/guest/test/-/gradebook/list/title/desc/20/false/1/assignment/master
> 
> (See the attached screenshot  as an example of what it looks like in a CRUD portlet)
> 
> mvcbean-gradebook-viewstate
> 
> Here are some *example usages*, relying heavily on the JSP Expression Language (EL) in order to attain a *purely declarative syntax* for a JSP view:
> 
> https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/webapp/WEB-INF/views/assignments.jspx#L14
> https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/webapp/WEB-INF/views/assignments.jspx#L30-L35

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @ngriffin7a, thanks for sending this and thanks @drewbrokke for forwarding it!

I've pasted the JIRA description here, but it's still a bit unclear to me:
- What are we trying to achieve with this? What's the problem it's solving?
- Where does this apply? What do you expect from this PR?

We're doing some heavy work around `DataSets` and the legacy tags, so this comes at an interesting time (as it gives us some additional insight on what might be needed) at the same time as inconvenient, since it grows the API to new lengths...

I'm tagging specially @brunobasto since he's probably the best one to deal with this PR, but hoping you can help clarify this a little bit.

Thanks!

**Comment by ngriffin7a** [comment]
Hi @brunobasto and @jbalsas -- I tried very hard to thoroughly document the problem, and how this PR solves the problem in the description field of the JIRA ticket: https://issues.liferay.com/browse/LPS-115019

The main goal is to cleanup JSPs by better supporting declarative views via the Java EE Expression Language (EL) feature:
https://docs.oracle.com/javaee/7/tutorial/jsf-el.htm

A secondary goal is to better support SearchContainer + FriendlyURL.

As I mentioned in the ticket, I have refactored the "Gradebook" portlet to use JSP declarative views. This is the portlet that students of our Liferay University training class end up building. Links to JSPs are found in the description of the JIRA ticket.

If you have any questions about the description I wrote in the ticket, then please let me know and I would be happy to answer. Thanks!

**Comment by brunobasto** [comment]
Hey @ngriffin7a @jbalsas This looks like an interesting solution for our `SearchContainer`. For our coming `DataSetDisplay`, it doesn't implement routing so it doesn't rely on url parameters to grab render state. Instead, "navigation" is done with asynchronous requests and most changes to the UI are not reflected on the URL. Maybe that's something we need to add to it.

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Hey @ngriffin7a, thanks again for your work on this!

> I tried very hard to thoroughly document the problem, and how this PR solves the problem in the description field of the JIRA ticket: https://issues.liferay.com/browse/LPS-115019

I know and I really appreciate it!! I know I'm probably asking you to dumb it down for me a little, but I'd love to fully understand how it all plays together. I'll work to catch up on this, so please, bear with me for a little bit 🙏 

> The main goal is to cleanup JSPs by better supporting declarative views via the Java EE Expression Language (EL) feature:
https://docs.oracle.com/javaee/7/tutorial/jsf-el.htm

So, most of these changes are necessary in order to do

```html
<clay:management-toolbar displayContext="${assignmentClayToolbarViewState}" itemsTotal="${assignmentCount}" searchContainerId="assignments" />
```

Some questions:
- Is Java EE Expression Language (EL) something we should support across the board? Should we apply this or similar patterns to our other JSP tags?
- What's the exact connection between these interfaces and classes and the Gradebook example? I'm assuming [AssignmentClayToolbarViewState](https://github.com/ngriffin7a/liferay-training-backend-dev/blob/62830be9001165e05a5396dee55deffddbf09794/wars/gradebook-mvcbean-web/src/main/java/com/liferay/training/gradebook/web/view/AssignmentClayToolbarViewState.java)? Is this the way to connect those? 

> A secondary goal is to better support SearchContainer + FriendlyURL.

To make sure I understand... does that mean that from [this](https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/resources/META-INF/friendly-url-routes.xml#L22)

```xml
<route>
    <pattern>/{displayStyle}/{orderByCol}/{orderByType}/{delta}/{resetCur}/{cur}</pattern>
    <overridden-parameter name="javax.portlet.render">assignment</overridden-parameter>
</route>
```

The system will populate the beans with those values from the url?

I'm happy to leave @brunobasto lead this since he's more experienced and quickly got the jist of it, but I'd love to understand what we're doing here and see if there's additional room for improvement here!

From the tests results, only baseline seems relevant, which should be easily fixable... and naming seems somewhat off, so we can work that out a little before pushing forward... @brunobasto, can you take care of it?

@david-truong, what do you think of this? should we keep an eye on this and add it to our general roadmap? We're doing lot of tag development, so if we need to add support for this, the sooner the better!

**Comment by wincent** [comment]
ping

ping: @brunobasto @david-truong

**Comment by brunobasto** [comment]
Thanks for the reminder @wincent 

> I'm happy to let @brunobasto lead this since he's more experienced and quickly got the jist of it, but I'd love to understand what we're doing here and see if there's additional room for improvement for the future!

I got the intuition, but I'd have to play with it to get a better sense of it. I'd be happy to work on this.

> Is Java EE Expression Language (EL) something we should support across the board? Should we apply this or similar patterns to our other JSP tags?

I do not have a strong opinion on this. To me this is kind of a syntactic sugar, and I've seen the usage of _EL_ as a common practice in GS and outside Liferay.

These days the pattern in portal seems to be:

```jsp
<foo:bar baz="<%= myDisplayContext.getBaz() %>" />
```

The friendly url thing I think it's a good pattern that we should use more in portal and what Neil introduces here is something that will serve as an incentive for that to be used more I think.

> From the tests results, only baseline seems relevant, which should be easily fixable... and naming seems somewhat off, so we can work that out a little before pushing forward... @brunobasto, can you take care of it?

Yes. I'd be happy to.
@jbalsas can you give me a hint of what you meant by "naming seems somewhat off"? I could start from there.

Unless @david-truong has any objections, I will pull this and try to get it merged.

**Comment by ngriffin7a** [comment]
@brunobasto Thanks for reviewing!

@jbalsas  *TL;DR*, the EL helps us make declarative JSPs that simply "declare" what the view is. It helps us keep markup in JSP files, and Java in .java files. EL does this by being the *glue* that binds the view to the model in an MVC application. The EL has been available to portlet developers since Liferay 4.3.4. This PR simply adds some conveniences for developers to have some nice EL syntax for the Clay management toolbar, the SearchContainer, and FriendlyURLs.

Regarding the Expression Language (EL), it was first introduced in Java EE 5 (JSP 2.1) via [JSR 245](https://jcp.org/en/jsr/detail?id=245), circa 2005.

I have been using EL syntax exclusively for declarative JSP and Facelets in Liferay portlets ever since around 2008.

In a traditional server-side MVC implementation, EL can be understood as the "glue" that binds the *V*iew to the *M*odel.

To answer your specific questions:

> So, most of these changes are necessary in order to do
```
<clay:management-toolbar displayContext="${assignmentClayToolbarViewState}" itemsTotal="${assignmentCount}" searchContainerId="assignments" />
```

It's not just lines like that for the Clay management toolbar, but it's also for expressions that bind the view to the model. For example, from the Gradebook training, here is an input text field that binds the `<input/>` declarative markup to the model:
```
<input id="${namespace}firstName" class="form-control" name="${namespace}firstName" type="text"     value="${applicant.firstName}"/>
```

In this example, `${applicant.firstName}` is simply the equivalent of `<%= applicant.getFirstName() %>`

You see, it is the presence of JSP `<%=` ... `%>` "scriptlet" syntax that causes problems in IDEs.

Another benefit of declarative JSP views is the `<jsp:` tags. This enables the developer to create JSP or Facelet document that is purely declarative -- parseable by an XML parser. For example:

```
<?xml version="1.0" encoding="UTF-8"?>
<jsp:root xmlns:aui="http://liferay.com/tld/aui"
		  xmlns:c="http://java.sun.com/jsp/jstl/core"
		  xmlns:fmt="http://java.sun.com/jsp/jstl/fmt"
		  xmlns:jsp="http://java.sun.com/JSP/Page"
		  xmlns:liferay-ui="http://liferay.com/tld/ui"
		  xmlns:portlet="http://xmlns.jcp.org/portlet_3_0"
		  version="2.1">
    ...
</jsp:root>
```

For a full document that shows all the features I just outlined, see:
https://github.com/ngriffin7a/liferay-blade-samples/blob/applicant/maven/apps/applicant/applicant-mvcbean-cdi-jsp-portlet/src/main/webapp/WEB-INF/views/applicant.jspx

> Is Java EE Expression Language (EL) something we should support across the board? Should we apply this or similar patterns to our other JSP tags?

The EL has been available for JSP portlet developers since Liferay 4.4.2 (circa 2008) when we moved to Tomcat 6 (Java EE 5, JSP 2.1):
http://tomcat.apache.org/whichversion.html

My pull request simply makes use of EL to be *more convenient* with the Clay management toolbar and SearchContainer.
Any request (typically a RenderRequest or ResourceRequest) or JSP PageContext attributes are automatically available for use in EL expressions. If you want to introduce additional stuff into the EL, that's easy, simply set a request attribute. For example:

```
renderRequest.setAttribute("assignment", new Assignment());
```

You can also introduce things into the EL if a JSP tag has a `var` attribute. For example:

```
<porlet:actionURL var="myActionURL" />
<form method="post" action="${myActionURL}">
</form>
```

Finally, CDI and/or Spring let's you use the `@Named` attribute to introduce a Java class instance into the EL. For example:

```
@PortletRequestScoped
@Named("myFoo")
public class Foo {
    ...
}
```

https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/java/com/liferay/training/gradebook/web/view/ClayToolbarViewStateProducer.java#L58

> What's the exact connection between these interfaces and classes and the Gradebook example? I'm assuming AssignmentClayToolbarViewState? Is this the way to connect those?

Yes! :-) The `ClayToolbarViewStateProducer` is a factory (using CDI which is familiar to the Java EE developer, but it can also be done with Declarative Services) that is responsible for creating a *request scoped* instance of an `AssignmentClayToolbarViewState` and uses `@Named` in order to introduce it into the EL:
https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/java/com/liferay/training/gradebook/web/view/ClayToolbarViewStateProducer.java#L58

But there is *also* the `SearchContainerViewStateProducer`, which is a factory that is responsible for creating a *request scoped* instance of a `SearchContainerViewState` and uses `@Named` in order to introduce a variable into the EL:
https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/java/com/liferay/training/gradebook/web/view/SearchContainerViewStateProducer.java#L39

> The system will populate the beans with those values from the url?

Yes! :-) The Liferay FriendlyURL mechanism relies on URL parameters, which makes the `javax.portlet.RenderURL` human readable, and bookmarkable.

**Comment by ngriffin7a** [comment]
@jbalsas Oh I forgot one more nice thing I wanted to mention. We can get rid of the `redirectURL` parameter! :-)

**Details:**

Portlet 3.0 introduces a new `type` attribute for `<portlet:param/>`

When this is used with a `<portlet:actionURL/>`, you can easily add "render" parameters that will survive from the `ACTION_PHASE` of the portlet lifecycle into the `RENDER_PHASE` so that those render parameters can be made available to JSP views that get rendered after an action. This is a GREAT matchup with `SearchContainer`, because it allows us to have a **STATELESS** way of passing along render state, and **ELIMINATE** the ugly `redirectURL` parameter that makes Liferay URLs so long and hard to read. Here is an example:

```
		<c:if test="${!readOnly}">
			<portlet:actionURL name="submitAssignment" var="formActionURL">
				<portlet:param name="javax.portlet.render" type="render" value="assignment" />
				<!-- SearchContainer view state -->
				<portlet:param name="cur" type="render" value="${searchContainerViewState.cur}" />
				<portlet:param name="delta" type="render" value="${searchContainerViewState.delta}" />
				<portlet:param name="displayStyle" type="render" value="${searchContainerViewState.displayStyle}" />
				<portlet:param name="orderByCol" type="render" value="${param.orderByCol}" />
				<portlet:param name="orderByType" type="render" value="${param.orderByType}" />
				<portlet:param name="resetCur" type="render" value="${searchContainerViewState.resetCur}" />
			</portlet:actionURL>
		</c:if>
```
See: https://github.com/ngriffin7a/liferay-training-backend-dev/blob/master/wars/gradebook-mvcbean-web/src/main/webapp/WEB-INF/views/assignment.jspx#L35-L40

**Comment by brunobasto** [comment]
Hey @ngriffin7a, perhaps you can enlighten me about something on the code. What are those Wrapper classes for?

**Comment by brunobasto** [comment]
NVM I see the usage here: https://github.com/ngriffin7a/liferay-training-backend-dev/blob/62830be9001165e05a5396dee55deffddbf09794/wars/gradebook-mvcbean-web/src/main/java/com/liferay/training/gradebook/web/view/AssignmentClayToolbarViewState.java#L34

**Comment by ngriffin7a** [comment]
@brunobasto

First, the wrapper pattern is very typical within Liferay's API, providing a convenient `@ConsumerType` that is designed to be forward-compatible, so that their code doesn't break on upgrades to newer versions of the API.

Second, the [ClayToolbarViewStateWrapper](https://github.com/jbalsas/liferay-portal/blob/pr-2219/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/view/state/ClayToolbarViewStateWrapper.java) class is there because it is better to **decorate** an instance of [ClayToolbarViewStateImpl](https://github.com/jbalsas/liferay-portal/blob/pr-2219/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/view/state/ClayToolbarViewStateImpl.java) than to `extend` it. Besides, it's an internal class, which means it is not part of the exposed/exported API.

The same can basically be said for `SearchContainerViewStateWrapper` and `SearchContainerViewStateImpl`.

**Comment by brunobasto** [comment]
Thanks for explanation, Neil

**Comment by brunobasto** [comment]
About the naming, It seems the pattern for abstract classes are to be named with `Base`. See `git grep "public abstract" "*.java"`.

Would  `BaseClayToolbarViewState` make sense?

**Comment by ngriffin7a** [comment]
@brunobasto I would rather remove the `abstract` keyword than prepend "Base" as a prefix and remove `"Wrapper"` as a suffix, so that the concept of the "Wrapper" pattern is clear from the class name.

FWIW, we added a whole bunch of "Wrapper" classes to the Portlet 3.0 API, and did not use the `abstract` keyword. Instead, we made the wrappers concrete. For example:
https://docs.liferay.com/portlet-api/3.0/javadocs/javax/portlet/filter/RenderURLWrapper.html

**Comment by brunobasto** [comment]
> @brunobasto I would rather remove the abstract keyword than prepend "Base" as a prefix and remove "Wrapper" as a suffix, so that the concept of the "Wrapper" pattern is clear from the class name.

That sounds good to me. I'm just trying to anticipate the stuff BChan usually questions me about.

**Comment by ngriffin7a** [comment]
@brunobasto Thanks for that. FWIW, I am the developer behind BeanPortletMethodWrapper.java and it cleared both Shuyang and BChan's review.

**Comment by brunobasto** [comment]
Yeah, I saw. Let's give it a try as is then :)

**Comment by brunobasto** [comment]
Just started reviewing :)

**Comment by brunobasto** [comment]
After sending https://github.com/liferay-frontend/liferay-portal/pull/84 I found out that the baseline update might no longer be necessary. So I'll trigger ci again and check..

**Comment by brunobasto** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Hey @brunobasto!

> @jbalsas can you give me a hint of what you meant by "naming seems somewhat off"? I could start from there.

There's 2 things that look off to me:
- We call the component `ManagementToolbar` but most of the classes somehow link to it as `ClayToolbar`. There's a bit of a mismatch there. 
- `SearchContainer` belongs in `liferay-ui`, which is currently `portal-web` and `util-taglib` (and even `portal-kernel` and `portal-impl`). It's hard for me to wrap my head around exposing these wrappers through `frontend-taglib-clay`. I'm assuming we can't do this from kernel and it needs to be an OSGi module, but then I'd probably prefer that we put this somewhere else to make it more obvious.

**Comment by jbalsas** [comment]
Hey @ngriffin7a, thanks a lot for those detailed and enlightening explanations!!

There's a lot of awesome material here that I think we would want to start promoting as best practices inside `liferay-portal`!!

Going to be taking notes and analyzing all the things you listed to see how we can benefit from it all!

@david-truong... still waiting on your input. I'd lean towards doing this in general for our tags and add this as part of our review/QA/acceptance criteria process for the future, but I'd love to get your input on how to prioritize this.

/cc @dsanz, interesting cross-cutting concern here 😉

**Comment by brunobasto** [comment]
> SearchContainer belongs in liferay-ui, which is currently portal-web and util-taglib (and even portal-kernel and portal-impl). It's hard for me to wrap my head around exposing these wrappers through frontend-taglib-clay. 

That didn't catch my attention until you mention. That is odd indeed. 

>  I'm assuming we can't do this from kernel and it needs to be an OSGi module, but then I'd probably prefer that we put this somewhere else to make it more obvious.

Yeah, I don't think we can put OSGi components outside `/modules`. Would it be appropriate for this module to live inside `modules/apps/liferay-frontend` or would it need its own new module entirely? Looking at the list under `/modules` nothing that would make sense immediately stands out as a good place for it.

**Comment by jbalsas** [comment]
> modules/apps/liferay-frontend

Do you mean somewhere in `modules/apps/frontend-js/` or `modules/apps/frontend-taglib` or is `liferay-frontend` a complete different thing? 

The `liferay-frontend:*` tags are located in `modules/apps/frontend-taglib/frontend-taglib`. If that's what you're referring, that was my initial thought.

I'm a tad bit worried about this becoming an API, because it means we have to remember to maintain and support and avoid breaking and we don't have that much expertise on this yet... but we'll have to bite the bullet :)

**Comment by brunobasto** [comment]
> Do you mean somewhere in modules/apps/frontend-js/ or modules/apps/frontend-taglib or is liferay-frontend a complete different thing?

Oops, I meant `modules/apps/frontend-taglib`. So my question is: Does it make sense for SearchContainerViewState to be in some submodule inside `modules/apps/frontend-taglib` given that in practive t's not namespaced with `liferay-frontend:`?

If we are to be strict about it, I  think it doesn't make sense. But then it would mean either:
1. We put `SearchContainerViewState` in a new module outside `frontend-taglib` and `ManagementToolbarViewState` inside `frontend-taglib` and we maintain both
2. We put `SearchContainerViewState` in a new module outside `frontend-taglib` and `ManagementToolbarViewState` inside `frontend-taglib` and we only own `ManagementToolbarViewState`.

Question: Do we own `liferay-ui:search-container`?

- I prefer 2 because it best follows our current module/ownership logic. We don't own TagLibs from WCM or DDM teams for example. Also, this is the kind of pattern that makes more sense for each TagLib implementers to follow as they are implementing them.

> I'm a tad bit worried about this becoming an API, because it means we have to remember to maintain and support and avoid breaking and we don't have that much expertise on this yet... but we'll have to bite the bullet :)

That is true.

**Comment by jbalsas** [comment]
> Question: Do we own liferay-ui:search-container?

**Yes!!**. We own all legacy tags, `liferay-ui:*`, `aui:*`, `liferay-theme:*`, `liferay-portlet:*`...

Makes sense for us to maintain both... at some point we'll want to move the `SearchContainer` thing to `DataSetDisplay` which might make it easier for us.

If we want to go ahead with this I'd suggest, for simplicity, try the following:
- Place `SearchContainer` stuff in `frontend-taglib`
- Keep `ManagementToolbar` stuff in `frontend-taglib-clay`. Not sure why you wouldn't actually want that there 🤔

**Comment by jbalsas** [comment]
So, just to confirm, the old `moono-lisa` doesn't seem to have that `.cke_editable` portion, so it won't affect AlloyEditor, but the new `moono-lexicon` does... 🤷

**Comment by julien** [comment]
> So, just to confirm, the old moono-lisa doesn't seem to have that .cke_editable portion, so it won't affect AlloyEditor, but the new moono-lexicon does... shrug

@jbalsas was this comment for #66?
(With @carloslancha we aslo think the issue here comes from the skin just like you said, but we'll talk about that in #66)

**Comment by brunobasto** [comment]
> Not sure why you wouldn't actually want that there 🤔

I wasn't proposing to put it anywhere else. Just didn't write the complete path `frontend-taglib-clay`, but yeah let's keep it in there

**Comment by jbalsas** [comment]
> I wasn't proposing to put it anywhere else. Just didn't write the complete path frontend-taglib-clay, but yeah let's keep it in there

Hahaha... I'm doing too many things at once today and you're messing with my head!! 😂

**Comment by brunobasto** [comment]
Ahh I see the confusion. We have `frontend-taglib/frontend-taglib`. Sorry bout that. Above whenever I mentioned `frontend-taglib` I meant `module/frontend-taglib-clay` and I was asking if we should place it outside `module/frontend-taglib`. Like `modules/search-container-view-state` or something.

**Comment by brunobasto** [comment]
to me it's fine to have it inside `modules/frontend-taglib/frontend-taglib`.

Just a reminder: It will be an exposed package, so we won't be able to move it later without it being a breaking change.

**Comment by brunobasto** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
> Just a reminder: It will be an exported package, so we won't be able to move it later without it being a breaking change.

I know... that's what scares me :)

> modules/search-container-view-state

potatotomato

I don't think it makes such a big difference... but if we do more of them it makes sense they're co-located with their tags, so maybe `frontend-taglib` (even if `search-container`) isn't there is the least of the evils...

**Comment by ngriffin7a** [comment]
My best understanding is that Shuyang is hesitant to add new modules unless it is absolutely necessary, in order to keep startup time as fast as possible.

**Comment by brunobasto** [comment]
Thanks, @ngriffin7a. That makes sense. I'll move it to `modules/apps/frontend-taglib/frontend-taglib` and push the new commit here.

**Comment by brunobasto** [comment]
@jbalsas `frontend-taglib` already depends on `frontend-taglib-clay`. Since `ManagementToolbarViewStateFactoryImpl` depends on `SearchContainerURLFactory`, if we move `SearchContainerURLFactory` to `frontend-taglib` that means `frontend-taglib-clay` needs tot depend on `frontend-taglib` as well. That creates a circular dependency :/

I'll look for alternatives.

**Comment by ngriffin7a** [comment]
To @brunobasto @jbalsas @drewbrokke and everyone on this thread:

I just wanted to say **THANK YOU** for putting so much time into this! It is very encouraging me to receive such wonderful support! :-)

**Comment by jbalsas** [comment]
> That creates a circular dependency :/

Sigh... I'd lean towards another module, then, even with the startup considerations I would give that a try... hopefully we can promise to remove some modules to make up for it. We've actually already removed some in this release, so we should have some credit! 😂 

> I just wanted to say THANK YOU for putting so much time into this! It is very encouraging me to receive such wonderful support! :-)

You're very welcome, @ngriffin7a! We're the ones that are grateful for your contributions and your patience explaining everything! ☺️

**Comment by ngriffin7a** [comment]
I think Shuyang will be OK with another module as long as it starts up dynamically, rather than statically.

**Comment by brunobasto** [comment]
How does https://github.com/liferay-frontend/liferay-portal/pull/71/commits/19819bbce7285ce69a401978f63a1333d8398271 look?

**Comment by brunobasto** [comment]
ci:test:sf

**Comment by brunobasto** [comment]
ci:test:sf

**Comment by brunobasto** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
shipit

**Comment by jbalsas** [comment]
ci:forward

**Comment by jbalsas** [comment]
ci:forward

**Comment by ngriffin7a** [comment]
See https://github.com/drewbrokke/liferay-portal/pull/1135
